### PR TITLE
feat: Search within date range in pipeline run API

### PR DESCRIPTION
**Pull request by yuechao-qin**
### TL;DR

Implemented search within date range in Pipeline Runs.

#### Functional

- API `GET /api/pipeline_runs/`
  - Search by date range in `filter_query`. Example query (not URL encoded for example):
    ```
      /api?filter_query={
        "and": [{
          "time_range": {
            "key": "system/pipeline_run.date.created_at",
            "start_time": "2024-01-01T00:00:00Z",
            "end_time": "2024-02-01T00:00:00Z"
          }
        }]
      }
    ```
  - Valid time ranges:
    - Start time only
    - End time only
    - (Start + End) time
    

#### Other

- Added timezone handling to convert aware datetimes to naive UTC for database compatibility (i.e. date range search will convert API timezone to UTC because DB datetime is only UTC)

### How to test?

```
uv run pytest tests/test_api_server_sql.py tests/test_filter_query_sql.py
```

The implementation includes comprehensive tests covering:
- Basic time range filtering with start and end times
- Start-only time ranges (no end time)
- Boundary conditions (inclusive start, exclusive end)
- Timezone offset handling (e.g., `+05:30` converted to UTC)
- Negation with `NOT` operator
- Combination with annotation-based filters
- Nested logical operations (`AND`/`OR`)
- Multiple time ranges in the same query

### Why make this change?

- This change enables users to filter pipeline runs by creation date range.
- The implementation properly handles timezone conversions to ensure consistent behavior across different client timezones while maintaining compatibility with the database's naive UTC storage format.

### CODE REVIEW

File: cloud_pipelines_backend/filter_query_sql.py [hunk 1]
```diff
@@ -361,3 +369,56 @@ def _value_in_to_clause(
             bts.PipelineRunAnnotation.value.in_(predicate.value_in.values),
         ],
     )
+
+
+# ---------------------------------------------------------------------------
```
Comment (yuechao-qin) [inline cloud_pipelines_backend/filter_query_sql.py:374]: Add tests for start_time is None.  Changes from code review.

File: cloud_pipelines_backend/filter_query_sql.py [hunk 2]
```diff
@@ -363,3 +369,59 @@ def _value_in_to_clause(
...
+      UTC-aware datetime
+        datetime(2024, 1, 1, 2, 30, 0, tzinfo=UTC)
+              |
+              v  .replace(tzinfo=None) -- strips timezone label
+      Naive datetime
+        datetime(2024, 1, 1, 2, 30, 0)
+              |
+              v  SQLAlchemy literal_binds -- adds microsecond precision
+      SQL string
+        '2024-01-01 02:30:00.000000'  <-- matches DB storage format
```
Comment (Ark-kun) [inline cloud_pipelines_backend/filter_query_sql.py:411]: I'm not sure this will work properly. We take naive timestamp from the user and convert it to certain timezone. I'm not sure the result is correct. We do not know the user's timezone. The backend timezone will be used (which one?).

If the user gives us naive timestamps, let's treat them as UTC.

File: cloud_pipelines_backend/filter_query_sql.py [hunk 3]
```diff
@@ -363,3 +369,59 @@ def _value_in_to_clause(
...
+            "time_range only supports key "
+            f"{PipelineRunAnnotationSystemKey.CREATED_AT!r}, got {tr.key!r}"
+        )
+    # Convert aware datetimes to naive UTC to match DB storage format.
+    clauses: list[sql.ColumnElement] = []
+    if tr.start_time is not None:
+        start_utc = tr.start_time.astimezone(datetime.timezone.utc).replace(tzinfo=None)
+        clauses.append(bts.PipelineRun.created_at >= start_utc)
+    if tr.end_time is not None:
+        end_utc = tr.end_time.astimezone(datetime.timezone.utc).replace(tzinfo=None)
```
Comment (Ark-kun) [inline cloud_pipelines_backend/filter_query_sql.py:425]: Just curious: What happens if we do not do `.replace(tzinfo=None)`

**Review by Ark-kun — CHANGES_REQUESTED** [review]
Let's make sure the timezones work well.

File: cloud_pipelines_backend/filter_query_sql.py (hunk 2 above)
Comment (yuechao-qin) [reply cloud_pipelines_backend/filter_query_sql.py:411]: I'm confused on this comment. Let me make sure we're aligned:

- My implementation will NOT ALLOW naive timestamps.  It requires timezone-aware timestamps only.  The Pydantic model uses `AwareDatetime` for start/end time paramters.  If the client does not provide a timezone (i.e. naive), Pydantic will return a 422 validation error.
- Thus the user is required to send use timezone-aware timestamps.  We convert it to UTC, since that's what the DB uses.
- Are you suggesting we allow naive timestamps?  I rather not do that, because naive timestamps are ambiguous, we can't know if the user meant UTC, their local timezone, or the server's timezone. Silently assuming UTC could give wrong results for users in other timezones. Requiring an explicit timezone removes all guesswork and makes the API contract unambiguous.

File: cloud_pipelines_backend/filter_query_sql.py (hunk 3 above)
Comment (yuechao-qin) [reply cloud_pipelines_backend/filter_query_sql.py:425]: If we did not do .replace(tzinfo=None), it will keep the timezone in the datetime used in the WHERE clause, which does not conform to the DB's naive datetime format. Depending on the DB, it'll react differently. Worst case is the datetime comparisons will be wrong.

File: cloud_pipelines_backend/filter_query_sql.py (hunk 2 above)
Comment (morgan-wowk) [reply cloud_pipelines_backend/filter_query_sql.py:411]: I believe it would be more straightforward if we just support UTC - that is probably what Alexey is suggesting.

The comment right now indicates that we support ISO 8601 format in general, which includes a variety of timezone offsets, not just UTC.

I think the solution you have appears like it would work well. It does introduce flexibility that we probably don't need; Yet, I don't imagine there would be a significant difference in lines of code. Just more careful testing.

The alternative would be to strictly enforce the format `2024-01-01T08:00:00Z` on the API rather than having to do the conversion because we support offsets.

Comment (Ark-kun) [reply cloud_pipelines_backend/filter_query_sql.py:411]: My main concern was with the `.astimezone(utc)`.  
For example, if timestamp is a naive UTC timestamp, then .astimezone(utc) will change it.  
  
\>If the client does not provide a timezone (i.e. naive), Pydantic will return a 422 validation error.  
  
OK. This lifts my concern.
